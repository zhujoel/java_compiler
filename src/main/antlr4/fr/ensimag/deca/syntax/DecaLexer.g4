lexer grammar DecaLexer;

options {
   language=Java;
   // Tell ANTLR to make the generated lexer class extend the
   // the named class, which is where any supporting code and
   // variables will be placed.
   superClass = AbstractDecaLexer;
}

@members {
}

// Deca lexer rules.

// Mots réservés
ASM : 'asm' ;
CLASS : 'class' ;
EXTENDS : 'extends' ;
ELSE : 'else' ;
FALSE : 'false' ;
IF : 'if' ;
INSTANCEOF : 'instanceof' ;
NEW  : 'new' ;
NULL : 'null' ;
READINT : 'readInt' ;
READFLOAT : 'readFloat' ;
PRINT : 'print' ;
PRINTLN : 'println'  ;
PRINTLNX : 'printlnx' ;
PRINTX : 'printx' ;
PROTECTED : 'protected' ;
RETURN : 'return' ;
THIS : 'this' ;
TRUE : 'true' ;
WHILE : 'while' ;

// Identificateurs
fragment LETTER : ('a' .. 'z') | ('A' .. 'Z');
fragment DIGIT : '0' .. '9' ;
IDENT : (LETTER | '$' | '_')(LETTER | DIGIT | '$' | '_')* ;

// Symboles spéciaux
LT : '<' ;
GT : '>' ;
EQUALS : '=' ;
PLUS : '+' ;
MINUS : '-' ;
TIMES : '*' ;
SLASH : '/' ;
PERCENT : '%' ;
DOT : '.' ;
COMMA : ',' ;
OPARENT : '(' ;
CPARENT : ')' ;
OBRACE : '{' ;
CBRACE : '}' ;
EXCLAM : '!' ;
SEMI : ';' ;
EQEQ : '==' ;
NEQ : '!=' ;
GEQ : '>=' ;
LEQ : '<=' ;
AND : '&&' ;
OR : '||' ;

// Littéraux entiers
fragment POSITIVE_DIGIT : '1' .. '9' ;
INT : '0' | POSITIVE_DIGIT DIGIT* ;

// Littéraux flottants
fragment NUM : DIGIT+ ;
fragment SIGN : (('+' | '-')?) ;
fragment EXP : ('E' | 'e') SIGN NUM ;
fragment DEC : NUM '.' NUM ;
fragment FLOATDEC : (DEC | DEC EXP) (('F' | 'f')?) ;
fragment HEX_MAJUSC : 'A' .. 'F' ;
fragment HEX_MINUSC : 'a' .. 'f' ;
fragment DIGITHEX : DIGIT | HEX_MAJUSC | HEX_MINUSC ;
fragment NUMHEX : DIGITHEX+ ;
fragment FLOATHEX : ('0x' | '0X') NUMHEX '.' NUMHEX ('P' | 'p') SIGN NUM (('F' | 'f' )?) ;
FLOAT : FLOATDEC | FLOATHEX ;

// Chaines de caractères
fragment STRING_CAR : (~('"' | '\\' | '\n'))+ ;
STRING : '"' (STRING_CAR | '\\"' | '\\\\')* '"' ;
MULTI_LINE_STRING : '"' (STRING_CAR | '\n' | '\\"' | '\\\\')* '"' ;

// Commentaires
MULTI_LINE_COMMENT : '/*' MULTI_LINE_STRING '*/' { skip() ; } ;
MONO_LINE_COMMENT : '//' (~('\n'))* ('\n' | EOF) { skip() ; } ;

// Séparateurs
SEPAR : ( ' ' | '\t' | '\n' | '\r' ) { skip(); } ;

// Inclusion de fichier
fragment FILENAME : (LETTER | DIGIT | '.' | '-' | '_')+ ;
INCLUDE : '#include' (' ')* '"' FILENAME '"' { doInclude(getText()); };
