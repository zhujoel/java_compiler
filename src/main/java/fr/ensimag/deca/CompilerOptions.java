package fr.ensimag.deca;

import fr.ensimag.ima.pseudocode.Register;
import java.io.File;
import java.io.PrintStream;
import java.text.ParseException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import org.apache.log4j.Level;
import org.apache.log4j.Logger;

/**
 * User-specified options influencing the compilation.
 *
 * @author gl48
 * @date 01/01/2020
 */
public class CompilerOptions {
    public static final int QUIET = 0;
    public static final int INFO  = 1;
    public static final int DEBUG = 2;
    public static final int TRACE = 3;
    public int getDebug() {
        return debug;
    }

    public boolean getParallel() {
        return parallel;
    }
    
    public boolean getVerification(){
        return verification;
    }

    public boolean getPrintBanner() {
        return printBanner;
    }

    public boolean getRegisters() {
        return registers;
    }

    public int getX() {
        return x;
    }

    public boolean getParse() {
        return parse;
    }

    public boolean getAllCompilation() {
        return allCompilation;
    }
    
    public List<File> getSourceFiles() {
        return Collections.unmodifiableList(sourceFiles);
    }

    private int debug = 0;
    private boolean parallel = false;
    private boolean printBanner = false;
    private boolean verification = false;
    private boolean noCheck = false;
    private boolean parse = false;
    private boolean registers = false;
    private boolean allCompilation = false;
    private int x;
    private List<File> sourceFiles = new ArrayList<File>();

    
    public void parseArgs(String[] args) throws CLIException {
        int i = 0;
        while (i < args.length) {
                switch (args[i]) {
                    //option pour faire toute la compilation et 
                    //generer et montrer le fichier .ass
                    case "-a":
                        allCompilation = true; 
                        break;
                    case "-b":
                        if (i == 0 && args.length == 1) {
                            printBanner = true;
                        } else {
                            System.out.println("The -b option cannot "
                                    + "be used with other options");
                            System.exit(1);
                        }
                        break;
                    case "-p":
                        if (verification) {
                            System.err.println("Les options [-p] et [-v] "
                                    + "sont incompatibles.");
                            System.exit(1);
                        } else {
                        parse = true;
                        System.out.println("Option -p ");
                        }
                        break;
                    case "-P":
                            parallel = true;
                            System.out.println("Option -P");
                        
                        break;
                    case "-d":
                        debug++;
                        break;
                    case "-v":
                        if (parse) {
                            System.err.println("Les options [-p] et [-v] "
                                    + "sont incompatibles.");
                            System.exit(1);
                        } else {
                            verification = true;
                            System.out.println("Option -v");
                        }
                        break;
                    case "-n":
                        noCheck = true;
                        break;
                    case "-r":
                        try {
                            x = Integer.parseInt(args[i + 1]);
                            if (x >= 4 && x <= 16) {// 4 <= X <= 16
                                System.out.println("X = " + x);
                                registers = true;
                                x -= 1;
                            } else {
                                System.out.println("X must have a value between 4 and 16");
                                System.exit(1);
                            }
                        } catch (NumberFormatException e) {
                            System.out.println("Invalid value for X");
                            System.exit(1);
                        } catch (ArrayIndexOutOfBoundsException a) {
                            System.out.println("No value for X was entered");
                            System.exit(1);
                        }
                        break;
                    default: sourceFiles.add(new File(args[i]));
                }
            
        
            i++;
        }
        
        if(sourceFiles.isEmpty() && args.length > 0 && !printBanner){
            System.err.println("No valid path for the source file was detected");
            System.exit(1);
        }else if(!sourceFiles.isEmpty() && args.length > 0 && parallel && sourceFiles.size() < 2){
            System.err.println("The [-P] option needs at least 2 source files");
            System.exit(1);
        }
        
        
        // DONE! A FAIRE : parcourir args pour positionner les options correctement.
        Logger logger = Logger.getRootLogger();
        // map command-line debug option to log4j's level.
        switch (getDebug()) {
            case QUIET:
                break; // keep default
            case INFO:
                logger.setLevel(Level.INFO);
                break;
            case DEBUG:
                logger.setLevel(Level.DEBUG);
                break;
            case TRACE:
                logger.setLevel(Level.TRACE);
                break;
            default:
                logger.setLevel(Level.ALL);
                break;
        }
        logger.info("Application-wide trace level set to " + logger.getLevel());

        boolean assertsEnabled = false;
        assert assertsEnabled = true; // Intentional side effect!!!
        if (assertsEnabled) {
            logger.info("Java assertions enabled");
        } else {
            logger.info("Java assertions disabled");
        }

    }

    protected void displayUsage(){//what should it do?
        //throw new UnsupportedOperationException("not yet implemented");
        System.out.println("");
    }
}

