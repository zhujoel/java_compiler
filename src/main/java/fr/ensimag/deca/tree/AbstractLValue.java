package fr.ensimag.deca.tree;

import fr.ensimag.deca.context.Type;
import fr.ensimag.deca.DecacCompiler;
import fr.ensimag.deca.context.ClassDefinition;
import fr.ensimag.deca.context.ContextualError;
import fr.ensimag.deca.context.EnvironmentExp;

/**
 * Left-hand side value of an assignment.
 * 
 * @author gl48
 * @date 01/01/2020
 */
public abstract class AbstractLValue extends AbstractExpr {
}
