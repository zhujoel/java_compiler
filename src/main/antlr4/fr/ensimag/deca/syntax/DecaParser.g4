parser grammar DecaParser;

options {
    // Default language but name it anyway
    //
    language  = Java;

    // Use a superclass to implement all helper
    // methods, instance variables and overrides
    // of ANTLR default methods, such as error
    // handling.
    //
    superClass = AbstractDecaParser;

    // Use the vocabulary generated by the accompanying
    // lexer. Maven knows how to work out the relationship
    // between the lexer and parser and will build the
    // lexer before the parser. It will also rebuild the
    // parser if the lexer changes.
    //
    tokenVocab = DecaLexer;

}

// which packages should be imported?
@header {
    import fr.ensimag.deca.tree.*;
    import java.io.PrintStream;
}

@members {
    @Override
    protected AbstractProgram parseProgram() {
        return prog().tree;
    }
}

prog returns[AbstractProgram tree]
    : list_classes main EOF {
            assert($list_classes.tree != null);
            assert($main.tree != null);
            $tree = new Program($list_classes.tree, $main.tree);
            setLocation($tree, $list_classes.start);
        }
    ;

main returns[AbstractMain tree]
    : /* epsilon */ {
            $tree = new EmptyMain();
        }
    | block {
            assert($block.decls != null);
            assert($block.insts != null);
            $tree = new Main($block.decls, $block.insts);
            setLocation($tree, $block.start);
        }
    ;

block returns[ListDeclVar decls, ListInst insts]
    : OBRACE list_decl list_inst CBRACE {
            assert($list_decl.tree != null);
            assert($list_inst.tree != null);
            $decls = $list_decl.tree;
            $insts = $list_inst.tree;
        }
    ;

list_decl returns[ListDeclVar tree]
@init   {
            $tree = new ListDeclVar();
        }
    : decl_var_set[$tree]*
    ;

decl_var_set[ListDeclVar l]
    : type list_decl_var[$l,$type.tree] SEMI
    ;

list_decl_var[ListDeclVar l, AbstractIdentifier t]
    : dv1=decl_var[$t] {
        $l.add($dv1.tree);
        } (COMMA dv2=decl_var[$t] {
        	// on ajoute dans la liste 
        	$l.add($dv2.tree);
        }
      )*
    ;

decl_var[AbstractIdentifier t] returns[AbstractDeclVar tree]
@init   {
        }
    : i=ident {
        }
      ((EQUALS e=expr {
      		Initialization init = new Initialization($e.tree);
      		$tree = new DeclVar($t, $i.tree, init);
      		setLocation(init, $e.start);
      		setLocation($tree, $EQUALS);
        }
      ) | {
			// nouvelle déclaration
			$tree = new DeclVar($t, $i.tree, new NoInitialization());
    		setLocation($tree, $i.start);
        })
    ;

list_inst returns[ListInst tree]
@init {
	$tree = new ListInst();
}
    : (inst {
    		$tree.add($inst.tree);
        }
      )*
    ;

inst returns[AbstractInst tree]
    : e1=expr SEMI {
            assert($e1.tree != null);
            $tree = $e1.tree;
            setLocation($tree, $e1.start);
        }
    | SEMI {
    		$tree = new NoOperation();
            setLocation($tree, $SEMI);
        }
    | PRINT OPARENT list_expr CPARENT SEMI {
            assert($list_expr.tree != null);
            $tree = new Print(false, $list_expr.tree);
            setLocation($tree, $PRINT);
        }
    | PRINTLN OPARENT list_expr CPARENT SEMI {
            assert($list_expr.tree != null);
            $tree = new Println(false, $list_expr.tree);
            setLocation($tree, $PRINTLN);
        }
    | PRINTX OPARENT list_expr CPARENT SEMI {
            assert($list_expr.tree != null);
            $tree = new Print(true, $list_expr.tree);
            setLocation($tree, $PRINTX);
        }
    | PRINTLNX OPARENT list_expr CPARENT SEMI {
            assert($list_expr.tree != null);
            $tree = new Println(true, $list_expr.tree);
            setLocation($tree, $PRINTLNX);
        }
    | if_then_else {
            assert($if_then_else.tree != null);
            $tree = $if_then_else.tree;
            setLocation($tree, $if_then_else.start);
        }
    | WHILE OPARENT condition=expr CPARENT OBRACE body=list_inst CBRACE {
            assert($condition.tree != null);
            assert($body.tree != null);
            $tree = new While($condition.tree, $body.tree);
            setLocation($tree, $WHILE);
        }
    | RETURN expr SEMI {
            assert($expr.tree != null);
            $tree = $expr.tree;
            setLocation($tree, $RETURN);
        }
    ;

if_then_else returns[IfThenElse tree]
@init {
	ListInst elseBranch = new ListInst();
	ListInst elseIfElseBranch = new ListInst();
	boolean hasIfElse = false;
}
    : if1=IF OPARENT condition=expr CPARENT OBRACE li_if=list_inst CBRACE {
    		$tree = new IfThenElse($condition.tree, $li_if.tree, elseBranch);
            setLocation($tree, $if1);
        }
      (ELSE elsif=IF OPARENT elsif_cond=expr CPARENT OBRACE elsif_li=list_inst CBRACE {
      		if(!hasIfElse){
        		elseBranch.add(new IfThenElse($elsif_cond.tree, $elsif_li.tree, elseIfElseBranch));
        		hasIfElse = true;
        	}
        	else{
        		// on ajoute les else if en + récursivement dans la boucle else du else if précédent
        		ListInst elseIfElseRecBranch = new ListInst();
        		elseIfElseBranch.add(new IfThenElse($elsif_cond.tree, $elsif_li.tree, elseIfElseRecBranch));
        		
        	}
        }
      )*
      (ELSE OBRACE li_else=list_inst CBRACE {
      		if(!elseBranch.isEmpty()){
      			for(AbstractInst i : $li_else.tree.getList()){
      				elseIfElseBranch.add(i);
      			}
      		}
      		else{
      			for(AbstractInst i : $li_else.tree.getList()){
      				elseBranch.add(i);
      			}
      		}
        }
      )?
    ;

list_expr returns[ListExpr tree]
@init   {
            $tree = new ListExpr();
        }
    : (e1=expr {
    		$tree.add($e1.tree);
        }
       (COMMA e2=expr {
       		$tree.add($e2.tree);
        }
       )* )?
    ;

expr returns[AbstractExpr tree]
    : assign_expr {
            assert($assign_expr.tree != null);
            $tree = $assign_expr.tree;
			setLocation($tree, $assign_expr.start);
        }
    ;

assign_expr returns[AbstractExpr tree]
    : e=or_expr (
        /* condition: expression e must be a "LVALUE" */ {
            if (! ($e.tree instanceof AbstractLValue)) {
                throw new InvalidLValue(this, $ctx);
            }
            $tree = $e.tree;
			setLocation($tree, $e.start);
        }
        EQUALS e2=assign_expr {
            assert($e.tree != null);
            assert($e2.tree != null);
			$tree = new Assign((AbstractLValue)$e.tree, $e2.tree);
			setLocation($tree, $EQUALS);
            
        }
      | /* epsilon */ {
            assert($e.tree != null);
            $tree = $e.tree;
        }
      )
    ;

or_expr returns[AbstractExpr tree]
    : e=and_expr {
            assert($e.tree != null);
            $tree = $e.tree;
			setLocation($tree, $e.start);
        }
    | e1=or_expr OR e2=and_expr {
            assert($e1.tree != null);
            assert($e2.tree != null);
            $tree = new Or($e1.tree, $e2.tree);
			setLocation($tree, $e1.start);
       }
    ;

and_expr returns[AbstractExpr tree]
    : e=eq_neq_expr {
            assert($e.tree != null);
            $tree = $e.tree;
			setLocation($tree, $e.start);
        }
    |  e1=and_expr AND e2=eq_neq_expr {
            assert($e1.tree != null);                         
            assert($e2.tree != null);
            $tree = new And($e1.tree, $e2.tree);
			setLocation($tree, $e1.start);
        }
    ;


eq_neq_expr returns[AbstractExpr tree]
    : e=inequality_expr {
            assert($e.tree != null);
            $tree = $e.tree;
			setLocation($tree, $e.start);
        }
    | e1=eq_neq_expr EQEQ e2=inequality_expr {
            assert($e1.tree != null);
            assert($e2.tree != null);
            $tree = new Equals($e1.tree, $e2.tree);
			setLocation($tree, $e1.start);
        }
    | e1=eq_neq_expr NEQ e2=inequality_expr {
            assert($e1.tree != null);
            assert($e2.tree != null);
            $tree = new NotEquals($e1.tree, $e2.tree);
			setLocation($tree, $e1.start);
        }
    ;


inequality_expr returns[AbstractExpr tree]
    : e=sum_expr {
            assert($e.tree != null);
            $tree = $e.tree;
			setLocation($tree, $e.start);
        }
    | e1=inequality_expr LEQ e2=sum_expr {
            assert($e1.tree != null);
            assert($e2.tree != null);
            $tree = new LowerOrEqual($e1.tree, $e2.tree);
			setLocation($tree, $e1.start);
        }
    | e1=inequality_expr GEQ e2=sum_expr {
            assert($e1.tree != null);
            assert($e2.tree != null);
            $tree = new GreaterOrEqual($e1.tree, $e2.tree);
			setLocation($tree, $e1.start);
        }
    | e1=inequality_expr GT e2=sum_expr {
            assert($e1.tree != null);
            assert($e2.tree != null);
            $tree = new Greater($e1.tree, $e2.tree);
			setLocation($tree, $e1.start);
        }
    | e1=inequality_expr LT e2=sum_expr {
            assert($e1.tree != null);
            assert($e2.tree != null);
            $tree = new Lower($e1.tree, $e2.tree);
			setLocation($tree, $e1.start);
        }
    | e1=inequality_expr INSTANCEOF type {
            assert($e1.tree != null);
            assert($type.tree != null);
			setLocation($tree, $e1.start);
        }
    ;



sum_expr returns[AbstractExpr tree]
    : e=mult_expr {
            assert($e.tree != null);
            $tree = $e.tree;
			setLocation($tree, $e.start);
        }
    | e1=sum_expr PLUS e2=mult_expr {
            assert($e1.tree != null);
            assert($e2.tree != null);
            $tree = new Plus($e1.tree, $e2.tree);
			setLocation($tree, $e1.start);
        }
    | e1=sum_expr MINUS e2=mult_expr {
            assert($e1.tree != null);
            assert($e2.tree != null);
            $tree = new Minus($e1.tree, $e2.tree);
			setLocation($tree, $e1.start);
        }
    ;

mult_expr returns[AbstractExpr tree]
    : e=unary_expr {
            assert($e.tree != null);
            $tree = $e.tree;
			setLocation($tree, $e.start);
        }
    | e1=mult_expr TIMES e2=unary_expr {
            assert($e1.tree != null);                                         
            assert($e2.tree != null);
            $tree = new Multiply($e1.tree, $e2.tree);
			setLocation($tree, $e1.start);
        }
    | e1=mult_expr SLASH e2=unary_expr {
            assert($e1.tree != null);                                         
            assert($e2.tree != null);
            $tree = new Divide($e1.tree, $e2.tree);
			setLocation($tree, $e1.start);
        }
    | e1=mult_expr PERCENT e2=unary_expr {
            assert($e1.tree != null);                                                                          
            assert($e2.tree != null);
            $tree = new Modulo($e1.tree, $e2.tree);
			setLocation($tree, $e1.start);
        }
    ;

unary_expr returns[AbstractExpr tree]
    : op=MINUS e=unary_expr {
            assert($e.tree != null);
            $tree = new UnaryMinus($e.tree);
			setLocation($tree, $op);
        }
    | op=EXCLAM e=unary_expr {
            assert($e.tree != null);
            $tree = new Not($e.tree);
			setLocation($tree, $op);
        }
    | select_expr {
            assert($select_expr.tree != null);
            $tree = $select_expr.tree;
			setLocation($tree, $select_expr.start);
        }
    ;

select_expr returns[AbstractExpr tree]
    : e=primary_expr {
            assert($e.tree != null);
            $tree = $e.tree;
			setLocation($tree, $e.start);
        }
    | e1=select_expr DOT i=ident {
            assert($e1.tree != null);
            assert($i.tree != null);
            // TODO: compléter
            
        }
        (o=OPARENT args=list_expr CPARENT {
            // we matched "e1.i(args)"
            assert($args.tree != null);
            // TODO: compléter
        }
        | /* epsilon */ {
            // we matched "e.i"
            // TODO: compléter
        }
        )
    ;

primary_expr returns[AbstractExpr tree]
    : ident {
            assert($ident.tree != null);
            $tree = $ident.tree;
			setLocation($tree, $ident.start);
        }
    | m=ident OPARENT args=list_expr CPARENT {
            assert($args.tree != null);
            assert($m.tree != null);
            // TODO: compléter
            
        }
    | OPARENT expr CPARENT {
            assert($expr.tree != null);
            $tree = $expr.tree;
			setLocation($tree, $OPARENT);
        }
    | READINT OPARENT CPARENT {
    		$tree = new ReadInt();
			setLocation($tree, $READINT);
        }
    | READFLOAT OPARENT CPARENT {
    		$tree = new ReadFloat();
			setLocation($tree, $READFLOAT);
        }
    | NEW ident OPARENT CPARENT {
            assert($ident.tree != null);
            $tree = $ident.tree;
			setLocation($tree, $NEW);
        }
    | cast=OPARENT type CPARENT OPARENT expr CPARENT {
            assert($type.tree != null);
            assert($expr.tree != null);
            // TODO: compléter
        }
    | literal {
            assert($literal.tree != null);
            $tree = $literal.tree;
			setLocation($tree, $literal.start);
        }
    ;

type returns[AbstractIdentifier tree]
    : ident {
            assert($ident.tree != null);
            $tree = $ident.tree;
			setLocation($tree, $ident.start);
        }
    ;

literal returns[AbstractExpr tree]
    : INT {
    		$tree = new IntLiteral(Integer.parseInt($INT.text));
			setLocation($tree, $INT);
        }
    | fd=FLOAT {
    		$tree = new FloatLiteral(Float.parseFloat($fd.text));
			setLocation($tree, $fd);
        }
    | STRING {
    		$tree = new StringLiteral($STRING.text);
			setLocation($tree, $STRING);
        }
    | TRUE {
    		$tree = new BooleanLiteral(true);
			setLocation($tree, $TRUE);
        }
    | FALSE {
    		$tree = new BooleanLiteral(false);
			setLocation($tree, $FALSE);
        }
    | THIS {
            // TODO: compléter
    		//if ($THIS.text.equals("this")){
    		//	$tree = new BooleanLiteral(false);
    		//}
    		//else{
    		//	$tree = new BooleanLiteral(true);
    		//}
        }
    | NULL {
            // TODO: compléter
    		//$tree = new NullLiteral();
    		//$tree.text = 'null';
    		}
    ;

ident returns[AbstractIdentifier tree]
    : IDENT {
    		// on crée un identifier
    		$tree = new Identifier(super.getDecacCompiler().getSymbolTable().create($IDENT.text));
			setLocation($tree, $IDENT);
        }
    ;

/****     Class related rules     ****/

// Liste de classes
list_classes returns[ListDeclClass tree]
@init{
	$tree = new ListDeclClass();
}
    :
      (c1=class_decl {
      		assert($c1.tree != null);
      		$tree.add($c1.tree);
        }
      )*
    ;

// Déclaration d'une classe
class_decl returns[AbstractDeclClass tree]
    : CLASS name=ident superclass=class_extension OBRACE class_body CBRACE {
            assert($ident.tree != null);
            assert($superclass.tree != null);
            assert($class_body.lfields != null);
            assert($class_body.lmethods != null);
    		$tree = new DeclClass($ident.tree, $superclass.tree, 
    			$class_body.lfields, $class_body.lmethods);
        }
    ;

// Extension d'une classe
class_extension returns[AbstractIdentifier tree]
    : EXTENDS ident {
    		assert($ident.tree != null);
    		// si la classe étend une autre
    		$tree = $ident.tree;
        }
    | /* epsilon */ {
    		// si elle étend rien, bah elle étend quand meme Object
    		$tree = new Identifier(super.getDecacCompiler().getSymbolTable().create("Object"));
        }
    ;

// Corps de la classe, avec des attributs et des méthodes
class_body returns[ListDeclField lfields, ListDeclMethod lmethods]
@init{
	$lmethods = new ListDeclMethod();
	$lfields = new ListDeclField();
}
    : (m=decl_method {
    		$lmethods.add($m.tree);
        }
      // passe la liste en paramètre pour que les fils le remplisse
      | decl_field_set[$lfields]
      )*
    ;

// Un attribut de la classe
decl_field_set[ListDeclField lfields]
    : v=visibility t=type list_decl_field[$lfields, $v.tree, $t.tree] SEMI
    ;

// visibilité d'un attribut ou méthode
visibility returns[Visibility tree]
    : /* epsilon */ {
    		// par défaut c'est public
    		$tree = Visibility.PUBLIC;
        }
    | PROTECTED {
    		$tree = Visibility.PROTECTED;
        }
    ;

// on ajoute l'attribut dans la liste d'attributs de la classe
list_decl_field[ListDeclField lfields, Visibility v, AbstractIdentifier t]
    : dv1=decl_field[$t, $v]{
    	$lfields.add($dv1.tree);
    }
        (COMMA dv2=decl_field[$t, $v]{
        	$lfields.add($dv2.tree);
        }
      )*
    ;

// déclaration de l'attribut de la classe avec initialisation ou non
decl_field[AbstractIdentifier t, Visibility v] returns[AbstractDeclField tree]
    : i=ident {
        }
       // si il y a une initialisation à l'attribut
      ((EQUALS e=expr {
      		Initialization init = new Initialization($e.tree);
      		$tree = new DeclField($v, $t, $i.tree, init);
        }
      )
      // pas d'initialisation 
      | {
      		$tree = new DeclField($v, $t, $i.tree, new NoInitialization());
        })
    ;

decl_method returns[AbstractDeclMethod tree]
@init {
	ListDeclParam params = new ListDeclParam();
}
	// déclaration d'une méthode
    : type ident OPARENT params=list_params[params] CPARENT (block {
    		$tree = new DeclMethod($type.tree, $ident.tree, params,
    			$block.decls, $block.insts);
        }
      |
      // TODO: faire l'ASM pour mettre de l'assembleur en tant que méthode de la classe 
      ASM OPARENT code=multi_line_string CPARENT SEMI {
        }
      ) {
        }
    ;

list_params[ListDeclParam params]
    : (p1=param {
    		params.add($p1.tree);
        } (COMMA p2=param {
        	params.add($p2.tree);
        }
      )*)?
    ;
    
multi_line_string returns[String text, Location location]
    : s=STRING {
            $text = $s.text;
            $location = tokenLocation($s);
        }
    | s=MULTI_LINE_STRING {
            $text = $s.text;
            $location = tokenLocation($s);
        }
    ;

param returns[AbstractDeclParam tree]
    : type ident {
    	$tree = new DeclParam($type.tree, $ident.tree);
        }
    ;