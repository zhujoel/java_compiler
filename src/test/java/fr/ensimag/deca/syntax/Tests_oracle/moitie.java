/* A manual test for the initial sketch of parsing included in 
 * students skeleton.
 * 
 */
package fr.ensimag.deca.syntax.Tests_oracle;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileOutputStream;
import java.io.FileReader;
import java.io.IOException;
import java.io.PrintStream;

import org.antlr.v4.runtime.CommonTokenStream;

import fr.ensimag.deca.CompilerOptions;
import fr.ensimag.deca.DecacCompiler;
import fr.ensimag.deca.syntax.AbstractDecaLexer;
import fr.ensimag.deca.syntax.DecaLexer;
import fr.ensimag.deca.syntax.DecaParser;
import fr.ensimag.deca.tree.AbstractExpr;
import fr.ensimag.deca.tree.AbstractIdentifier;
import fr.ensimag.deca.tree.AbstractLValue;
import fr.ensimag.deca.tree.AbstractProgram;
import fr.ensimag.deca.tree.Assign;
import fr.ensimag.deca.tree.DeclVar;
import fr.ensimag.deca.tree.FloatLiteral;
import fr.ensimag.deca.tree.Identifier;
import fr.ensimag.deca.tree.ListDeclClass;
import fr.ensimag.deca.tree.ListDeclVar;
import fr.ensimag.deca.tree.ListExpr;
import fr.ensimag.deca.tree.ListInst;
import fr.ensimag.deca.tree.Location;
import fr.ensimag.deca.tree.Main;
import fr.ensimag.deca.tree.Multiply;
import fr.ensimag.deca.tree.NoInitialization;
import fr.ensimag.deca.tree.Println;
import fr.ensimag.deca.tree.Program;
import fr.ensimag.deca.tree.ReadInt;
import fr.ensimag.deca.tree.Tree;
import fr.ensimag.deca.tree.TreeFunction;

public class moitie {
	
	private static DecacCompiler compiler = new DecacCompiler(null,null);
	
	static String currentUsersDir = System.getProperty("user.dir");
    
    /* script pour le test de moitie.deca */
    public static AbstractProgram ProgInit() {
        ListInst linst = new ListInst();
        ListDeclVar lDecl = new ListDeclVar();
        AbstractProgram source =
            new Program(
                new ListDeclClass(),
                new Main(lDecl,linst));
        
        AbstractIdentifier type = new Identifier(compiler.getSymbolTable().create("int"));
        AbstractIdentifier varName = new Identifier(compiler.getSymbolTable().create("x"));
        NoInitialization init = new NoInitialization();
        lDecl.add(new DeclVar(type, varName, init));
        
        AbstractLValue left = new Identifier(compiler.getSymbolTable().create("x"));
        AbstractExpr right = new ReadInt();
        linst.add(new Assign(left, right));
        ListExpr list_expr = new ListExpr();
        AbstractExpr carre = new Multiply(varName, new Identifier(compiler.getSymbolTable().create("x")));
        System.out.println("hello");
        AbstractExpr f = new FloatLiteral((float) 0.5);
        AbstractExpr expr = new Multiply(f, carre);
        list_expr.add(expr);
        linst.add(new Println(false, list_expr));
		
        return source;
    }
    
    public static void deleteLocation(Tree tree) {
    	Location location = null;
    	tree.setLocation(location);
    }
    
    protected static void deleteAllLocations(AbstractProgram prog) {
    	prog.iter(new TreeFunction() {
    		@Override
    		public void apply(Tree t) {
    			deleteLocation(t);
    		}
    	});
    }
    
    public static void genSyntaxTreeParserFile(String[] sourceName) throws IOException {
    	DecaLexer lex = AbstractDecaLexer.createLexerFromArgs(sourceName);
        CommonTokenStream tokens = new CommonTokenStream(lex);
        DecaParser parser = new DecaParser(tokens);
        File file = null;
        if (lex.getSourceName() != null) {
            file = new File(lex.getSourceName());
        }
        final DecacCompiler decacCompiler = new DecacCompiler(new CompilerOptions(), file);
        parser.setDecacCompiler(decacCompiler);
        AbstractProgram prog = parser.parseProgramAndManageErrors(System.err);
        deleteAllLocations(prog);
        if (prog == null) {
            System.exit(1);
        } else {
        	PrintStream ps = new PrintStream(new FileOutputStream(currentUsersDir + "/obtained/moitie.txt" , true));
            prog.prettyPrint(ps);
            ps.close();
        }
    }
    
    public static void genSyntaxTreeManualFile(AbstractProgram source) throws IOException {        
        PrintStream ps = new PrintStream(new FileOutputStream(currentUsersDir + "/expected/moitie.txt", true));
        source.prettyPrint(ps);
        ps.close();
        
    }

    public static void main(String args[]) throws IOException {
    	
    	int userspathlength = currentUsersDir.length();
    	String[] fichier_teste = new String[1];
    	String path = currentUsersDir.substring(0, userspathlength - 41) + "/deca/syntax/valid/created/moitie.deca";
        fichier_teste[0] = path;
        
        BufferedReader in = new BufferedReader(new FileReader(fichier_teste[0]));
		String line;
		while ((line = in.readLine()) != null)
		{
	      // Afficher le contenu du fichier
			  System.out.println (line);
		}
		in.close();
		AbstractProgram source = ProgInit();

        genSyntaxTreeManualFile(source);

        genSyntaxTreeParserFile(fichier_teste);
    }
        
   
}


