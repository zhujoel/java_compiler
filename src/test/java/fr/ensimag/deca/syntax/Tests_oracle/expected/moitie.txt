`> Program
   +> ListDeclClass [List with 0 elements]
   `> Main
      +> ListDeclVar [List with 1 elements]
      |  []> DeclVar
      |      +> Identifier (int)
      |      +> Identifier (x)
      |      `> NoInitialization
      `> ListInst [List with 2 elements]
         []> Assign
         ||  +> Identifier (x)
         ||  `> ReadInt
         []> Println
             `> ListExpr [List with 1 elements]
                []> Multiply
                    +> FloatLiteral (0.5)
                    `> Multiply
                       +> Identifier (x)
                       `> Identifier (x)
